### ISSUE TITLE
Can't query enum fields

### ISSUE THREAD

**Issue by lukejagodzinski** [samdenty/gqless#84]
I've discovered that in `gqless` it's not possible to query fields of enum type. Here is reproduction https://codesandbox.io/s/gqless-enum-bug-4lu4b

In this reproduction for each story I'm also querying the `type` field which is of enum type.
The resulting query should be:

```gql
{
  hn {
    topStories(limit: 25) {
      id
      type
      url
      title
      score
      by {
        id
      }
      descendants
    }
  }
}

```

but instead it tries to query:

```gql
{
  hn {
    topStories(limit: 25) {
      id
      type {
        __typename
      }
      url
      title
      score
      by {
        id
      }
      descendants
    }
  }
}
```

The error message is: `"Field "type" of type "ItemType!" must not have a sub selection.`
And you can see that it treats the enum type like object type.

Also, speaking about enums. It would be great if enums after generation are still enums and not type in the form: 

```ts
type t_ItemType = EnumType<"job" | "story" | "comment" | "poll" | "pollopt">;
```

but this instead:

```ts
enum t_ItemType {
  job = "job",
  story = "story",
  comment = "comment",
  poll = "poll",
  pollopt = "pollopt"
}
```

It's what tools like `graphql-codegen` do and it's pretty handy as I can use syntax like `ItemType.job` instead of providing value directly.

**Comment by Yomanz** [comment]
I started implementing this but there doesn't seem to be a general type for `enum`, so I'm a bit stuck.

**Comment by samdenty** [comment]
https://github.com/samdenty/gqless/blob/393e3f3510216ec130f99b137658a968ec78d5da/gqless/src/QueryBuilder/buildSelection.ts#L22

**Comment by lukejagodzinski** [comment]
Yep I'm just investigating and I got to this point :) I'm trying to understand how query generation works under the hood. Any suggestion so I could create PR?

**Comment by lukejagodzinski** [comment]
@samdenty I guess the only thing that has to be done is to treat `EnumNode`the same way as `ScalarNode`. It's kinda like scalar value and it can't have nested field. So here https://github.com/samdenty/gqless/blob/393e3f3510216ec130f99b137658a968ec78d5da/gqless/src/QueryBuilder/buildSelection.ts#L18 it should be:

```ts
if (innerNode instanceof ScalarNode || innerNode instanceof EnumNode) return ''
```

**Comment by samdenty** [comment]
Yeah was just about to say, that should fix it

**Comment by lukejagodzinski** [comment]
Ok, I will prepare PR if you're ok with that?

**Comment by lukejagodzinski** [comment]
@samdenty also what do you think about exposing enum types in the form of real enums? I know that it will probably mess with TS types as `EnumType` extends `Type` type. But maybe there is some way of achieving it? Just asking if you might consider that. It's not a big deal though however would be nice to have

**Comment by samdenty** [comment]
@lukejagodzinski Open to changing it, but will probably need to investigate. Maybe just changin the `export type X = TypeData` to `export enum X {}` will be enough

**Comment by samdenty** [comment]
Using the updated version, seems to still be broken. Think it's EnumNode#getData is broken https://codesandbox.io/s/gqless-enum-bug-qyzxt

**Comment by samdenty** [comment]
https://github.com/samdenty/gqless/blob/64e0505022207f5989b471e2fca67076d4575961/gqless/src/Cache/merge/merge.ts#L23

**Comment by Yomanz** [comment]
> 
> 
> @samdenty also what do you think about exposing enum types in the form of real enums? I know that it will probably mess with TS types as `EnumType` extends `Type` type. But maybe there is some way of achieving it? Just asking if you might consider that. It's not a big deal though however would be nice to have

I'm currently implementing that.

**Comment by lukejagodzinski** [comment]
@samdenty ok I will further investigate that. Have to better understand how the whole package works :). I will work on that and let you know.

@Yomanz that's great! Thanks!

**Comment by Yomanz** [comment]
Once this is merged you can install the latest package and use the proper enum types :)

https://github.com/samdenty/gqless/pull/87

**Comment by lukejagodzinski** [comment]
@samdenty @Yomanz I've prepared another PR now I've checked if it's working on the repository's example and it does. So sorry for previous not complete PR.

@samdenty thanks for all your suggestions. I've also modified the `onDataChange.ts` file. I don't understand everything yet about this package but my guess is that if on data change we check scalar values then we should also check enum values as well. I've also checked all the occurences of ScalarNode instance checks and that seems to be all. Let me know if it still requires some changes.

I would also write tests for this changes but running tests breaks on my computer, I mean like most of them even without my changes. Is there some special way of running tests here?

**Comment by samdenty** [comment]
Awesome! Thanks for the PR :)

> I would also write tests for this changes
Haven't tested much of the gqless packages yet. Also somehow broke tests, so no worries

**Comment by lukejagodzinski** [comment]
@samdenty ok cool :)

**Comment by lukejagodzinski** [comment]
@samdenty there is one problem with latest release, it doesn't work :P. I was investigating why for last couple of minutes and I found out that not all changes were released. I guess it's because I didn't add changeset to the second PR. I've assumed that the first one will be enough (I've never used changesets before so wasn't familiar with that). So I guess I should add changeset everytime I make a PR. I don't know what to do about it know, to no break it even more :)

Just to assure you, the change itself is working properly. When I'm building current version locally and link that in my test repo everything works properly. So it's just what was released that is a problem

**Comment by Yomanz** [comment]
> 
> 
> @samdenty there is one problem with latest release, it doesn't work :P. I was investigating why for the last couple of minutes and I found out that not all changes were released. I guess it's because I didn't add changeset to the second PR. I've assumed that the first one will be enough (I've never used changesets before so wasn't familiar with that). So I guess I should add changeset everytime I make a PR. I don't know what to do about it now, to no break it even more :)
> 
> Just to assure you, the change itself is working properly. When I'm building current version locally and link that in my test repo everything works properly. So it's just what was released that is a problem

Send a link to the PR you're referring to.

**Comment by lukejagodzinski** [comment]
@Yomanz this one https://github.com/samdenty/gqless/pull/88

**Comment by Yomanz** [comment]
> 
> 
> @Yomanz this one #88

I'll add it to the changeset of the PR I'm currently making.

**Comment by lukejagodzinski** [comment]
> > @Yomanz this one #88
> 
> I'll add it to the changeset of the PR I'm currently making.

@Yomanz ok great thanks!
